### OCI Data Science - Useful Tips
<details>
<summary><font size="2">Check for Public Internet Access</font></summary>

```python
import requests
response = requests.get("https://oracle.com")
assert response.status_code==200, "Internet connection failed"
```
</details>
<details>
<summary><font size="2">Helpful Documentation </font></summary>
<ul><li><a href="https://docs.cloud.oracle.com/en-us/iaas/data-science/using/data-science.htm">Data Science Service Documentation</a></li>
<li><a href="https://docs.cloud.oracle.com/iaas/tools/ads-sdk/latest/index.html">ADS documentation</a></li>
</ul>
</details>
<details>
<summary><font size="2">Typical Cell Imports and Settings for ADS</font></summary>

```python
%load_ext autoreload
%autoreload 2
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

import logging
logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.ERROR)

import ads
from ads.dataset.factory import DatasetFactory
from ads.automl.provider import OracleAutoMLProvider
from ads.automl.driver import AutoML
from ads.evaluations.evaluator import ADSEvaluator
from ads.common.data import ADSData
from ads.explanations.explainer import ADSExplainer
from ads.explanations.mlx_global_explainer import MLXGlobalExplainer
from ads.explanations.mlx_local_explainer import MLXLocalExplainer
from ads.catalog.model import ModelCatalog
from ads.common.model_artifact import ModelArtifact
```
</details>
<details>
<summary><font size="2">Useful Environment Variables</font></summary>

```python
import os
print(os.environ["NB_SESSION_COMPARTMENT_OCID"])
print(os.environ["PROJECT_OCID"])
print(os.environ["USER_OCID"])
print(os.environ["TENANCY_OCID"])
print(os.environ["NB_REGION"])
```
</details>

In [53]:
import pandas as pd

# Bank Marketing

## 1) Business Understanding
### Objective:
#### The classification goal is to predict if the client will subscribe a term deposit

### Key Questions:
#### What are the target market segment for the Portuguese banking institution, and what shoud be customer retention policies

### Success Criteria:
#### Provide options to the Portuguese banking institution to optimize its marketing strategies, reduce costs, and increase the conversion rates of their marketing campaigns, thereby improving overall profitability

## 2) Data Understanding

## 2.1 Data Collection

The data is related with direct marketing campaigns of a Portuguese banking institution. 
   The marketing campaigns were based on phone calls. Often, more than one contact to the same client was required, 
   in order to access if the product (bank term deposit) would be (or not) subscribed.

In [81]:
from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
bank_marketing = fetch_ucirepo(id=222) 
  
# data (as pandas dataframes) 
X = bank_marketing.data.features 
y = bank_marketing.data.targets 
  
# variable information 
bank_marketing.variables

name     role         type      demographic  \
0           age  Feature      Integer              Age   
1           job  Feature  Categorical       Occupation   
2       marital  Feature  Categorical   Marital Status   
3     education  Feature  Categorical  Education Level   
4       default  Feature       Binary             None   
5       balance  Feature      Integer             None   
6       housing  Feature       Binary             None   
7          loan  Feature       Binary             None   
8       contact  Feature  Categorical             None   
9   day_of_week  Feature         Date             None   
10        month  Feature         Date             None   
11     duration  Feature      Integer             None   
12     campaign  Feature      Integer             None   
13        pdays  Feature      Integer             None   
14     previous  Feature      Integer             None   
15     poutcome  Feature  Categorical             None   
16            y   Target       Binary             None   

                                          description  units missing_values  
0                                                None   None             no  
1   type of job (categorical: 'admin.','blue-colla...   None             no  
2   marital status (categorical: 'divorced','marri...   None             no  
3   (categorical: 'basic.4y','basic.6y','basic.9y'...   None             no  
4                              has credit in default?   None             no  
5                              average yearly balance  euros             no  
6                                   has housing loan?   None             no  
7                                  has personal loan?   None             no  
8   contact communication type (categorical: 'cell...   None            yes  
9                        last contact day of the week   None             no  
10  last contact month of year (categorical: 'jan'...   None             no  
11   last contact duration, in seconds (numeric). ...   None             no  
12  number of contacts performed during this campa...   None             no  
13  number of days that passed by after the client...   None            yes  
14  number of contacts performed before this campa...   None             no  
15  outcome of the previous marketing campaign (ca...   None            yes  
16          has the client subscribed a term deposit?   None             no

In [30]:
# metadata 
bank_marketing.metadata

{'uci_id': 222,
 'name': 'Bank Marketing',
 'repository_url': 'https://archive.ics.uci.edu/dataset/222/bank+marketing',
 'data_url': 'https://archive.ics.uci.edu/static/public/222/data.csv',
 'abstract': 'The data is related with direct marketing campaigns (phone calls) of a Portuguese banking institution. The classification goal is to predict if the client will subscribe a term deposit (variable y).',
 'area': 'Business',
 'tasks': ['Classification'],
 'characteristics': ['Multivariate'],
 'num_instances': 45211,
 'num_features': 16,
 'feature_types': ['Categorical', 'Integer'],
 'demographics': ['Age', 'Occupation', 'Marital Status', 'Education Level'],
 'target_col': ['y'],
 'index_col': None,
 'has_missing_values': 'yes',
 'missing_values_symbol': 'NaN',
 'year_of_dataset_creation': 2014,
 'last_updated': 'Fri Aug 18 2023',
 'dataset_doi': '10.24432/C5K306',
 'creators': ['S. Moro', 'P. Rita', 'P. Cortez'],
 'intro_paper': {'title': 'A data-driven approach to predict the success of

In [82]:

X.head()

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married        NaN      no     1506     yes   no   
4   33           NaN   single        NaN      no        1      no   no   

  contact  day_of_week month  duration  campaign  pdays  previous poutcome  
0     NaN            5   may       261         1     -1         0      NaN  
1     NaN            5   may       151         1     -1         0      NaN  
2     NaN            5   may        76         1     -1         0      NaN  
3     NaN            5   may        92         1     -1         0      NaN  
4     NaN            5   may       198         1     -1         0      NaN

In [83]:
X.describe()

age        balance   day_of_week      duration      campaign  \
count  45211.000000   45211.000000  45211.000000  45211.000000  45211.000000   
mean      40.936210    1362.272058     15.806419    258.163080      2.763841   
std       10.618762    3044.765829      8.322476    257.527812      3.098021   
min       18.000000   -8019.000000      1.000000      0.000000      1.000000   
25%       33.000000      72.000000      8.000000    103.000000      1.000000   
50%       39.000000     448.000000     16.000000    180.000000      2.000000   
75%       48.000000    1428.000000     21.000000    319.000000      3.000000   
max       95.000000  102127.000000     31.000000   4918.000000     63.000000   

              pdays      previous  
count  45211.000000  45211.000000  
mean      40.197828      0.580323  
std      100.128746      2.303441  
min       -1.000000      0.000000  
25%       -1.000000      0.000000  
50%       -1.000000      0.000000  
75%       -1.000000      0.000000  
max      871.000000    275.000000

Total number of dataset : 45,211

Total number of features : 16

Target variable : has the client subscribed a term deposit? (binary: "yes","no")

## 2.2 Exploratory Data Analysis (EDA)

In [84]:
# To make data exploration easie for the first iteration, I am just loading 10,00 random rows into a dataframe

customer_sample = X.sample(n=10000)

# Explore the dataset
print(customer_sample.head())
print(customer_sample.info())
print(customer_sample.describe())


       age         job   marital  education default  balance housing loan  \
42696   64     retired   married  secondary      no      588      no   no   
36570   37  technician   married   tertiary      no     1733     yes   no   
18685   45  technician  divorced  secondary      no       72     yes   no   
18528   36  management   married   tertiary      no     3770     yes   no   
19937   44  management   married   tertiary      no     5581      no   no   

        contact  day_of_week month  duration  campaign  pdays  previous  \
42696  cellular           18   jan       366         1     91         2   
36570  cellular           12   may       524         3    326         1   
18685  cellular           31   jul       115         5     -1         0   
18528  cellular           31   jul       150         4     -1         0   
19937  cellular            8   aug       202         2     -1         0   

      poutcome  
42696  failure  
36570    other  
18685      NaN  
18528      NaN  
1

## 2.2 Handle categorical variables

In [86]:
 # Convert categorical variables into numerical values using one-hot encoding
    
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Define the categorical features
categorical_features = X.select_dtypes(include=['object']).columns

# One-hot encode categorical features
preprocessor = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(), categorical_features)
    ],
    remainder='passthrough'
)

# Fit and transform the data
X_encoded = preprocessor.fit_transform(X)

# Convert to DataFrame for easy interpretation (optional)
encoded_df = pd.DataFrame(X_encoded, columns=preprocessor.get_feature_names_out())
encoded_df['target'] = y

encoded_df

cat__job_admin.  cat__job_blue-collar  cat__job_entrepreneur  \
0                  0.0                   0.0                    0.0   
1                  0.0                   0.0                    0.0   
2                  0.0                   0.0                    1.0   
3                  0.0                   1.0                    0.0   
4                  0.0                   0.0                    0.0   
...                ...                   ...                    ...   
45206              0.0                   0.0                    0.0   
45207              0.0                   0.0                    0.0   
45208              0.0                   0.0                    0.0   
45209              0.0                   1.0                    0.0   
45210              0.0                   0.0                    1.0   

       cat__job_housemaid  cat__job_management  cat__job_retired  \
0                     0.0                  1.0               0.0   
1                     0.0                  0.0               0.0   
2                     0.0                  0.0               0.0   
3                     0.0                  0.0               0.0   
4                     0.0                  0.0               0.0   
...                   ...                  ...               ...   
45206                 0.0                  0.0               0.0   
45207                 0.0                  0.0               1.0   
45208                 0.0                  0.0               1.0   
45209                 0.0                  0.0               0.0   
45210                 0.0                  0.0               0.0   

       cat__job_self-employed  cat__job_services  cat__job_student  \
0                         0.0                0.0               0.0   
1                         0.0                0.0               0.0   
2                         0.0                0.0               0.0   
3                         0.0                0.0               0.0   
4                         0.0                0.0               0.0   
...                       ...                ...               ...   
45206                     0.0                0.0               0.0   
45207                     0.0                0.0               0.0   
45208                     0.0                0.0               0.0   
45209                     0.0                0.0               0.0   
45210                     0.0                0.0               0.0   

       cat__job_technician  ...  cat__poutcome_success  cat__poutcome_nan  \
0                      0.0  ...                    0.0                1.0   
1                      1.0  ...                    0.0                1.0   
2                      0.0  ...                    0.0                1.0   
3                      0.0  ...                    0.0                1.0   
4                      0.0  ...                    0.0                1.0   
...                    ...  ...                    ...                ...   
45206                  1.0  ...                    0.0                1.0   
45207                  0.0  ...                    0.0                1.0   
45208                  0.0  ...                    1.0                0.0   
45209                  0.0  ...                    0.0                1.0   
45210                  0.0  ...                    0.0                0.0   

       remainder__age  remainder__balance  remainder__day_of_week  \
0                58.0              2143.0                     5.0   
1                44.0                29.0                     5.0   
2                33.0                 2.0                     5.0   
3                47.0              1506.0                     5.0   
4                33.0                 1.0                     5.0   
...               ...                 ...                     ...   
45206            51.0               825.0                    17.0   
45207            71

### Split the data:

In [89]:
# Split the data
X = encoded_df.drop('target', axis=1)
y = encoded_df['target']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


Scale the features

In [95]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)



array([[-0.35938299,  1.90613718, -0.18348505, ..., -0.56588599,
        -0.41136376, -0.24477164],
       [-0.35938299, -0.52462121, -0.18348505, ..., -0.24538858,
        -0.41136376, -0.24477164],
       [-0.35938299, -0.52462121, -0.18348505, ..., -0.56588599,
        -0.41136376, -0.24477164],
       ...,
       [ 2.78254687, -0.52462121, -0.18348505, ..., -0.56588599,
        -0.41136376, -0.24477164],
       [ 2.78254687, -0.52462121, -0.18348505, ..., -0.24538858,
        -0.41136376, -0.24477164],
       [-0.35938299, -0.52462121, -0.18348505, ..., -0.24538858,
        -0.41136376, -0.24477164]])

## 3. Train and Evaluate Models

### Train each model and evaluate their performance using accuracy, precision, recall, F1-score, and ROC-AUC.

In [101]:
# Create a pipeline that includes both the preprocessor and the KNN model
pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', KNeighborsClassifier(n_neighbors=5))
])

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(encoded_df, encoded_df['target'], test_size=0.2, random_state=42)

# Fit the pipeline to the training data
pipeline.fit(X_train, y_train)

ValueError: A given column is not a column of the dataframe